File: Top 100 Codes/06-Advanced Codes related to Arrays/25-Smallest Subarray with sum greater than a given value/SmallestSubarraywithsumgreaterthanagivenvalue.py
def smallest_subarray_with_sum(arr, x):
    n = len(arr)
    min_len = float('inf')
    start = 0
    curr_sum = 0

    for end in range(n):
        curr_sum += arr[end]

        # Shrink the window from the start as long as the sum is greater than x
        while curr_sum > x:
            min_len = min(min_len, end - start + 1)
            curr_sum -= arr[start]
            start += 1

    return 0 if min_len == float('inf') else min_len

# Example usage
arr = [1, 4, 45, 6, 0, 19]
x = 51
print("Smallest subarray length:", smallest_subarray_with_sum(arr, x))

File: Top 100 Codes/06-Advanced Codes related to Arrays/25-Smallest Subarray with sum greater than a given value/SmallestSubarraywithsumgreaterthanagivenvalue.cpp
#include <iostream>
#include <vector>
#include <climits>
using namespace std;

int smallestSubarrayWithSum(vector<int>& arr, int x) {
    int n = arr.size();
    int minLen = INT_MAX;
    int start = 0, end = 0, currSum = 0;

    while (end < n) {
        // Expand window by including arr[end]
        currSum += arr[end];

        // Shrink window from the start as long as the sum is greater than x
        while (currSum > x) {
            minLen = min(minLen, end - start + 1);
            currSum -= arr[start];
            start++;
        }

        end++;
    }

    return (minLen == INT_MAX) ? 0 : minLen;
}

int main() {
    vector<int> arr = {1, 4, 45, 6, 0, 19};
    int x = 51;
    cout << "Smallest subarray length: " << smallestSubarrayWithSum(arr, x);
    return 0;
}
